File: day9/day9.c
#include <stdio.h>
#include <stdlib.h>

static size_t game(size_t marbles, size_t players)
{
	marbles++;
	size_t *next = malloc(marbles * sizeof(next[0]));
	size_t *prev = malloc(marbles * sizeof(prev[0]));
	size_t *score = calloc(players, sizeof(score[0]));
	if (!next || !prev || !score)
	{
		free(score);
		free(prev);
		free(next);
		return 0;
	}
	next[0] = prev[0] = 0;
	size_t cur = 0;
	size_t player = 0;
	for (size_t i = 1; i < marbles; i++)
	{
		if (i % 23 == 0)
		{
			size_t pos = cur;
			for (int j = 0; j < 7; j++)
			{
				pos = prev[pos];
			}
			next[prev[pos]] = next[pos];
			prev[next[pos]] = prev[pos];

			player %= players;
			score[player] += pos + i;
			cur = next[pos];
		}
		else
		{
			size_t pos1 = next[cur];
			size_t pos2 = next[pos1];
			next[i] = pos2;
			prev[i] = pos1;
			next[pos1] = i;
			prev[pos2] = i;
			cur = i;
		}
		player++;
	}
	size_t max = 0;
	for (size_t i = 0; i < players; i++)
	{
		if (max < score[i])
		{
			max = score[i];
		}
	}
	free(score);
	free(prev);
	free(next);
	return max;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s filename\n", argv[0]);
		return 1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "Cannot open %s\n", argv[1]);
		return 1;
	}

	size_t lastmarble, numplayers;
	int r = fscanf(input, " %zu players; last marble is worth %zu points", &numplayers, &lastmarble);
	fclose(input);
	if (r != 2)
	{
		fprintf(stderr, "Cannot parse the data\n");
		return 1;
	}

	printf("Part1: %zu\n", game(lastmarble, numplayers));
	printf("Part2: %zu\n", game(lastmarble*100U, numplayers));
	return 0;
}

File: day9/day9.py
#!/usr/bin/env python3

import sys

def game(marbles, players):
    marbles += 1
    score = [0] * players
    nxt = [0] * marbles
    prev = [0] * marbles
    cur = 0
    player = 0
    for i in range(1, marbles):
        if i % 23 == 0:
            pos = cur
            for j in range(7):
                pos = prev[pos]

            nxt[prev[pos]] = nxt[pos]
            prev[nxt[pos]] = prev[pos]

            player %= players
            score[player] += pos + i

            cur = nxt[pos]
        else:
            pos1 = nxt[cur]
            pos2 = nxt[pos1]

            nxt[i] = pos2
            prev[i] = pos1
            nxt[pos1] = i
            prev[pos2] = i

            cur = i

        player += 1

    return max(score)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "r") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    ar = txt.split(" ")
    num_players = int(ar[0])
    last_marble = int(ar[6])

    print("Part1:", game(last_marble, num_players))
    print("Part2:", game(last_marble*100, num_players))
